# 张量类

在第一部分，我们帮助小明构建了一个两层神经网络模型，可以根据天气预报来预测冰激凌销量。

在这个过程中，我们学习了多层神经网络的核心概念：

* **模型参数**：权重和偏置；
* **模型逻辑**：前向传播、梯度计算和反向传播；
* **超参数**：学习率、批大小和训练轮数。

我们也看到，随着网络层数的增加，前向传播、梯度计算和反向传播都要延展成一条更长的链路。我们需要手动维护这三条链路的顺序和完整性。可以想见，随着网络层数继续增加，手工维护会越发困难，也越发容易出错。

---

在这一部分，我们将尝试构建一套**神经网络训练框架**，把这三条链路的管理工作自动化。本章我们从最基础的数据结构开始：**张量**。

``💡 当前流行的神经网络训练框架主要分为两类：以 PyTorch 为代表的动态计算图框架，和以 TensorFlow 为代表的静态计算图框架。我们将采用动态计算图的方式，构建自己的神经网络训练框架。``

In [1]:
import numpy as np

## 基础架构

### 张量

**张量**（Tensor）是我们框架最基础的数据结构。

在数学上，张量是标量、向量和矩阵的统称与推广：单独一个数是标量（0 维张量）；一组数排成一行是向量（1 维张量）；数排成行列是矩阵（2 维张量）；更高维度的数组统称为张量。在深度学习里，"张量"通常就是这种 **N 维数组**的代称。

在我们的框架里，张量是对 NumPy 数组的**包装**。在张量内部，我们依然使用 NumPy 完成实际的数据存储和数值计算。那为什么不直接用 NumPy 数组，而要多此一举地包装它呢？

因为我们后续要在张量上附加更多能力。最重要的一项，是自动计算梯度的能力，也就是**自动微分**（Autograd）。这些是 NumPy 数组本身没有的功能，包装成张量类，让我们有地方来存放这些额外的状态和逻辑。

张量将是**梯度计算链路**的主要载体。

In [2]:
class Tensor:

    def __init__(self, data):
        self.data = np.array(data)

    def __repr__(self):
        return f'Tensor({self.data})'

目前的张量非常简单，只是把数据存进了 `self.data`。后续章节我们会逐步为它添加新能力。

## 数据

### 特征

在新的框架里，所有数据都将以张量的形式提供。我们来创建一个包含温度和湿度两个特征值的张量。

In [3]:
feature = Tensor([28.1, 58.0])
print(feature)

Tensor([28.1 58. ])


## 模型

### 参数：权重、偏置

所有模型参数也将以张量的形式存储。这里我们初始化一个神经元的权重和偏置。

In [4]:
weight = Tensor(np.ones((1, 2)) / 2)
bias = Tensor(np.zeros(1))
print(f'weight: {weight}')
print(f'bias:   {bias}')

weight: Tensor([[0.5 0.5]])
bias:   Tensor([0.])


### 推理函数（前向传播）

推理函数接收特征值张量和参数张量，完成线性计算，并以张量形式返回预测值。

In [5]:
def forward(x, w, b):
    return Tensor(x.data @ w.data.T + b.data)

## 验证

### 推理

In [6]:
prediction = forward(feature, weight, bias)
print(f'prediction: {prediction}')

prediction: Tensor([43.05])


预测值是 `43.05`。

至此，我们完成了框架的第一步：用张量统一描述所有数据。这看起来只是简单的包装，但它为后续自动微分打下了基础。在接下来的章节里，我们会逐步扩展张量的能力，自动完成梯度计算。

## 课后练习

用张量保存一个单独的数值（标量），例如 `Tensor(3.14)`，打印出来看看结果如何？